Сделано 1 задание

Krayseer opened the pull request.

Krayseer committed:
Добавлены сервисы для авторизации, обработки введённых данных с консоли и сохранения пользователей в файловую БД

Krayseer committed:
Убрал лишний вызов метода сохранения пользователей в БД из AuthenticationManager. Метод login() теперь void. Добавлен JavaDoc для CommandNames

Krayseer committed:
Исправил баг с несохранением данных после выключения приложения

Krayseer committed:
1) Для создания ApplicationProperties теперь нужно передать путь, где находятся настройки
2) Убран излишний класс AuthenticationManager, методы переписаны в AuthenticationService
3) Добавлен класс UserRepository для удобной манипуляции данных в БД. Для его реализации был создан интерфейс Repository
4) Убран класс CommandNames с константами существующих команд в приложении -> добавлен enum Commands с той же функцией, что и CommandNames
5) В список команд добавлена команда /help, с демонстрацией всех доступных команд
6) В классе ConsoleService в методе readCommand добавлена валидация на входные данные
7) Класс User перемещен в пакет domain
8) Добавлены тесты для классов AuthenticationService, CommandProcessor, ConsoleService, UserRepository

Krayseer committed:
1) В классе AuthenticationService исправлена логика работы метода logoutUser(), убрана лишняя проверка в authenticate()
2) Во всех классах убрано внедрение ConsoleService через аргумент конструктора
3) Проверка на существование команды перемещена из класса ConsoleService в CommandProcessor
4) Убран лишний интерфейс Repository
5) В ConsoleService в методе readUserFromConsole() добавлена валидация на ввод пустых значений
6) Обновлены тесты, добавлен утилитарный класс SteamUtils

Krayseer committed:
1) В перечисление команд добавлено свойство, которое сообщает о том, нужна ли для обработки команда аутентификация
2) Валидация команд на проверку, нужна ли аутентификация, добавлена в CommandProcessor в метод processCommand
3) Для создания UserRepository теперь нужно передать в аргументы путь до файловой БД
4) Добавлен интерфейс FileDBRepository с методами, для репозиториев, которые взаимодествуют с файлами БД
5) В тестах вместо физических файлов теперь используются временные файлы

Avrooid committed:
Добавил новые методы в ConsoleService для работы с пользователем, добавил сущности Contact и Group, добавил классы ContactRepository и GroupRepository для работы с файловой БД, добавил классы ContactService и GroupService, отвечающие за логику, добавил тесты для репозиториев ContactRepository и GroupRepository

Avrooid committed:
Добавил JavaDoc для констант в ContactService и GroupService, оптимизировал импорты, переделал BufferedWriter под FileUtils в ContactRepository и GroupRepository

vpyzhyanov commented on src/main/java/ru/anykeyers/commands/CommandProcessor.java:
```diff
@@ -0,0 +1,104 @@
+        commandHandlers.put(HELP, consoleService::writeCommands);
+        commandHandlers.put(EXIT_APP, () -> {
+            List<FileDBRepository> repositories = List.of(userRepository, contactRepository, groupRepository);
+            repositories.forEach(FileDBRepository::saveAll);
```
Если остановить приложение не через команду exit, то ничего не сохранится

Krayseer replied:
исправили на динамическое обновление данных после изменения данных какого нибудь Entity

vpyzhyanov commented on src/main/java/ru/anykeyers/commands/CommandHandler.java:
```diff
@@ -0,0 +1,13 @@
+    /**
+     * Обработать команду
+     */
+    void handleCommand();
```
Подозрительно почему этот метод ничего не принимает и не возвращает. Что тогда он обрабатывает? А как же параметры? А как же результат?

vpyzhyanov replied:
Похоже нашёл проблему. Если уже начал выполнять какую-то команду, то невозможно ввести другую команду в этот момент. Пока не завершишь операцию, ты находишься в западне. А должна быть возможность прервать операцию в любой момент.

vpyzhyanov replied:
Возможно ещё надо подумать насчёт этого замечания. С одной стороны - в консоли это допустимо. А с другой стороны это противоречит принципам работы чат-бот платформы. То есть если вы заходите перенести ваш проект на телеграмм (или любую другую чат платформу), то ничего у вас не получится.

Krayseer replied:
В обновленной версии возвращает String (результат обработки)

vpyzhyanov commented on src/main/java/ru/anykeyers/services/ContactService.java:
```diff
@@ -0,0 +1,144 @@
+        if (userContacts != null) {
+            for (Contact contact : userContacts) {
+                String tempName = contact.getFirstname() + " " + contact.getLastname();
+                if (tempName.equals(name)) {
```
Какова сложность этого метода? Спойлер: недопустимо большая. А вообще эту работу должен делать contactRepository

vpyzhyanov replied:
нет ответа! =(

Krayseer replied:
сложность метода будет O(n), в обновленной версии этого кода уже нет

vpyzhyanov commented on src/main/java/ru/anykeyers/services/ContactService.java:
```diff
@@ -0,0 +1,144 @@
+     * Добавляет контакт в список контактов пользователя
+     */
+    public void addContact() {
+        try (Scanner scanner = new Scanner(System.in)) {
```
Добро пожаловать на тёмную сторону ООП, у вас "коллективная безответственность". Про какие классы я говорю?

vpyzhyanov replied:
нет ответа! =(

Krayseer replied:
В этой версии(очень старой) код действительно очень сырой, вместо того чтобы добавлять какому нибудь пользователю контакт, мы в самом методе добавления контакта только считываем его, уже исправлено

vpyzhyanov commented on src/main/java/ru/anykeyers/services/ContactService.java:
```diff
@@ -0,0 +1,144 @@
+     * Удаляет контакт из списка контактов пользователя
+     */
+    public void deleteContact() {
+        try (Scanner scanner = new Scanner(System.in)) {
```
аналогично

Krayseer replied:
В этой версии(очень старой) код действительно очень сырой, вместо того чтобы добавлять какому нибудь пользователю контакт, мы в самом методе добавления контакта только считываем его, уже исправлено

vpyzhyanov commented on src/main/java/ru/anykeyers/services/ContactService.java:
```diff
@@ -0,0 +1,144 @@
+    public void deleteContact() {
+        try (Scanner scanner = new Scanner(System.in)) {
+            User user = authenticationService.getCurrentUser();
+            String nameContactToDelete = consoleService.deleteContactFromConsole(scanner);
```
как можно удалить контакт **из консоли**? Консоль - это средство вывода информации, а не хранения

Krayseer replied:
Действительно некорректное поведение, далее будет исправлено

vpyzhyanov commented on src/main/java/ru/anykeyers/services/ContactService.java:
```diff
@@ -0,0 +1,144 @@
+            Contact contactToDelete = contactRepository.findContactByNameOrNumber(user, nameContactToDelete);
+            if (contactToDelete != null) {
+                contactRepository.removeContact(user, contactToDelete);
+                System.out.println("Контакт удален");
```
ещё одна коллективная безотстветственность. В чём же дело тут?

vpyzhyanov replied:
нет ответа! =(

vpyzhyanov replied:
Всё ещё нет ответа( Если вы не будете отвечать на мои вопросы, то я не буду проверять ваши задачи!

Krayseer replied:
В методе удаления контакта мы сначала считываем контакт из консоли - это ошибка, лучше сразу в аргументы метода передавать контакт, который нужно удалить. в обновленных версиях этого нет

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/UserRepository.java:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+
+    public String getDbFilePath() {
```
Публичный метод? Во первых вообще непонятно почему путь до файла должен показываться всем кому попало. А во вторых вы нарушили инкапсуляцию, и главное вообще непонятно зачем.

Krayseer replied:
действительно плохое решение, в обновленных версиях убрано

vpyzhyanov commented on src/test/java/ru/anykeyers/repositories/UserRepositoryTest.java:
```diff
@@ -0,0 +1,91 @@
+        userRepository.save(user2);
+        userRepository.saveAll();
+
+        File dbFile = new File(userRepository.getDbFilePath());
```
И ради этого стоило уничтожить инкапсуляцию? Вы же сами задали этот путь при инициализации теста! У вас он уже есть, вы им управляете! Тогда зачем это было делать?!

Krayseer replied:
убрали этот фнкционал

vpyzhyanov commented on src/test/java/ru/anykeyers/repositories/UserRepositoryTest.java:
```diff
@@ -0,0 +1,91 @@
+
+        userRepository.save(user1);
+        userRepository.save(user2);
+        userRepository.saveAll();
```
Вот тут очень наглядно, что имя метода `saveAll` выбрано неудачно. Если читать код, мы прочитаем:

1. Сохранить user1
2. Сохранить user2
3. Сохранить всё..... стоп.... что всё? Мы же ничего не передали!? Куда тогда это сохранится? А предыдущие команды тогда куда сохраняют, если сохраняет только последняя?!

Krayseer replied:
функционал убран и добавлено динамическое обновление бд

vpyzhyanov commented on src/test/java/ru/anykeyers/repositories/UserRepositoryTest.java:
```diff
@@ -0,0 +1,91 @@
+
+        File dbFile = new File(userRepository.getDbFilePath());
+        assertTrue(dbFile.exists());
+        List<String> lines = FileUtils.readLines(dbFile, "UTF-8");
```
Лучше вызывать `org.apache.commons.io.FileUtils#readFileToString(java.io.File, java.nio.charset.Charset)` или ещё проще `Files.readString(dbFile.toPath(), StandardCharsets.UTF_8)` и проверять весь файл на equals сразу. Сделать всё одной проверкой, а не тремя.

Krayseer replied:
так и сделали в обновленной версии

vpyzhyanov reviewed (changes requested):
продолжение следует

Krayseer committed:
Исправлен подход к построению приложения
Теперь пользователь может сам выбрать какую БД ему использовать, и какое приложение создавать(телеграм бот, консольное приложение..) Реализовано это с помощью фабрик RepositoryFactory и ReceiverFactory
В классе User убран пароль из-за ненадобности
В Command добавлены параметры, чтобы производить валидацию перед обработкой команды на количество поступаемых аргументов
Добавлен файл messages.properties, чтобы иметь доступ к сообщениям приложения из одного файла по ключу
Исправлены тесты, убраны лишние классы(StreamUtils)
Теперь обработчик команд отдает результат обработки команды

Avrooid committed:
Добавил логику для Contact и Group: добавление, удаление, модификация; Поправил тесты для репозиториев и сервисов

Krayseer committed:
Добавлены тесты на класс CommandProcessor

Krayseer committed:
Изменен подход к обработке сообщений.
1) Теперь используется не ввод аргументов в строке, а обработка сообщений в зависимости от их типа, является ли сообщение командой, или сообщением, которое нужно обрабатывать в зависимости от состояния пользовтаеля
2) Добавлено обновление данных сразу в файловую БД после любых манипуляций для данных, что дает возможность работать из разных ботов
3) Добавлена реализация телеграм бота, теперь можно протестировать приложение сразу с несколькими запущенными ботами (консольный бот, телеграмм бот)

vpyzhyanov commented on src/main/java/ru/anykeyers/commands/Command.java:
```diff
@@ -0,0 +1,116 @@
+        return isNeedAuthenticate;
+    }
+
+    public static Command getCommandByValue(String value) {
```
static запрещён

vpyzhyanov replied:
не исправлено!

Krayseer replied:
Исправлено в новой версии

vpyzhyanov commented on src/main/java/ru/anykeyers/commands/Command.java:
```diff
@@ -0,0 +1,116 @@
+        return null;
+    }
+
+    public static String getAllCommands() {
```
static запрещён

vpyzhyanov replied:
не исправлено!

Krayseer replied:
исправлено в новой версии

vpyzhyanov commented on src/main/java/ru/anykeyers/ContacterApplication.java:
```diff
@@ -0,0 +1,56 @@
+        while (true) {
+            String command = receiver.readCommand();
+            String resultMessage = commandProcessor.processCommand(command);
+            receiver.sendMessage(resultMessage);
```
каким образом "**Получатель**" (именно так переводится слово receiver) может **отправить** сообщение? Он наоборот принимает сообщения. Пожалуйста, правильно выбирайте названия.

Krayseer replied:
в обновленной версии у нас это Bot)

vpyzhyanov commented on src/main/java/ru/anykeyers/receivers/Receiver.java:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Интерфейс, описывающий обработчик команд
+ */
+public interface Receiver {
```
Название и описание не соответствуют друг другу. Да и обработчик команд у вас уже есть, это `CommandProcessor`

Krayseer replied:
В обновленной версии будет исправлено

vpyzhyanov commented on src/main/java/ru/anykeyers/factories/ReceiverFactory.java:
```diff
@@ -0,0 +1,50 @@
+import ru.anykeyers.receivers.Receiver;
+import ru.anykeyers.receivers.impl.ConsoleReceiver;
+
+public class ReceiverFactory {
```
Нет JavaDoc

Krayseer replied:
В обновленной версии добавлено)

vpyzhyanov commented on src/main/java/ru/anykeyers/factories/ReceiverFactory.java:
```diff
@@ -0,0 +1,50 @@
+        /**
+         * Телеграмм бот
+         */
+        TELEGRAM_BOT,
```
Нигде не используется. Значит должно быть удалено

Krayseer replied:
Используется в обновленной версии

vpyzhyanov commented on src/main/java/ru/anykeyers/factories/ReceiverFactory.java:
```diff
@@ -0,0 +1,50 @@
+        /**
+         * Дискорд бот
+         */
+        DISCORD_BOT
```
Нигде не используется. Значит должно быть удалено

Krayseer replied:
Удалено

vpyzhyanov commented on src/main/java/ru/anykeyers/services/AuthenticationService.java:
```diff
@@ -0,0 +1,61 @@
+    /**
+     * Текущий авторизованный пользователь.
+     */
+    private User currentUser;
```
этот сервис будет работать **только** в консоли. В телеграмме и других чат-бот системах ничего работать не будет. Как думаете, почему?

Krayseer replied:
Многопользовательский режим, в обновленной версии с реализацией ботов это уже убрано (реализовано иначе)

vpyzhyanov commented on src/main/java/ru/anykeyers/factories/ReceiverFactory.java:
```diff
@@ -0,0 +1,50 @@
+            case CONSOLE:
+                return new ConsoleReceiver();
+            default:
+                throw new IllegalArgumentException();
```
добавьте человекочитаемое сообщение об ошибке

Krayseer replied:
Добавили в новой версии

vpyzhyanov replied:
Не исправлено!

vpyzhyanov replied:
> Добавили в новой версии

Наглая ложь!
![image](https://github.com/Krayseer/contacter/assets/71599019/2c2e9d9a-ea06-4287-81f0-a65b6b5a3386)

Krayseer replied:
Новая версия это имелось ввиду что я на тот момент уже делал изменения и конкретно сегодня и всех их запушил, и там уже это все исправлено

vpyzhyanov reviewed (changes requested):
Есть замечания

vpyzhyanov commented on src/main/java/ru/anykeyers/processors/commands/Command.java:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * Получить все команды
+     */
+    public static String getAllCommands() {
```
static запрещён!

Krayseer replied:
Исправлено в новой версии

vpyzhyanov commented on src/main/java/ru/anykeyers/processors/states/impl/ContactStateProcessor.java:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * Класс, содержащий информацию о действиях, которые можно сделать с контактом
+     */
+    private static final class ContactEditField {
```
static запрещён!

Krayseer replied:
Думал, что в объявлении внутреннего класса это не запрещено, поправим

vpyzhyanov commented on src/main/java/ru/anykeyers/processors/states/impl/GroupStateProcessor.java:
```diff
@@ -0,0 +1,101 @@
+    /**
+     * Класс, содержащий информацию о действиях, которые можно сделать с группой
+     */
+    private static final class GroupKind {
```
static запрещён!

Krayseer replied:
Думал, что в объявлении внутреннего класса это не запрещено

vpyzhyanov commented on src/main/java/ru/anykeyers/bots/Bot.java:
```diff
@@ -0,0 +1,26 @@
+     * @param username имя пользователя, которому нужно обработать сообщение
+     * @param message сообщение
+     */
+    void receiveMessage(String username, String message);
```
Не надо создавать никому не нужных методов в интерфейсе. Этот метод нигде не используется и должен быть удалён. Используются только **реализации** в конкретных классах, а сам интерфейс не используется нигде.

Krayseer replied:
Исправим

vpyzhyanov commented on src/main/java/ru/anykeyers/bots/Bot.java:
```diff
@@ -0,0 +1,26 @@
+    /**
+     * Запуск бота
+     */
+    void start();
```
хоть этот метод и используется, но его тоже можно легко удалить и ничего не изменится, т.к. по сути этот метод не нужен

Krayseer replied:
Хорошо, исправим

Krayseer commented:
этот недочет исправлен, в новой версии этого уже нет

сб, 2 дек. 2023 г. в 23:55, Пыжьянов Вячеслав ***@***.***>:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/main/java/ru/anykeyers/commands/Command.java
> <https://github.com/Krayseer/contacter/pull/4#discussion_r1412853554>:
>
> > +    }
> +
> +    public boolean isNeedAuthenticate() {
> +        return isNeedAuthenticate;
> +    }
> +
> +    public static Command getCommandByValue(String value) {
> +        for (Command command : Command.values()) {
> +            if (command.getCommandValue().equals(value)) {
> +                return command;
> +            }
> +        }
> +        return null;
> +    }
> +
> +    public static String getAllCommands() {
>
> не исправлено!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Krayseer/contacter/pull/4#discussion_r1412853554>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXT6LQO27QEVMSSFTGMNQSLYHN2SHAVCNFSM6AAAAAA7ICSTSCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTONRQHEZTSMRRGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Krayseer commented:
недочёт исправлен в новой версии

сб, 2 дек. 2023 г. в 23:55, Пыжьянов Вячеслав ***@***.***>:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/main/java/ru/anykeyers/commands/Command.java
> <https://github.com/Krayseer/contacter/pull/4#discussion_r1412853554>:
>
> > +    }
> +
> +    public boolean isNeedAuthenticate() {
> +        return isNeedAuthenticate;
> +    }
> +
> +    public static Command getCommandByValue(String value) {
> +        for (Command command : Command.values()) {
> +            if (command.getCommandValue().equals(value)) {
> +                return command;
> +            }
> +        }
> +        return null;
> +    }
> +
> +    public static String getAllCommands() {
>
> не исправлено!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Krayseer/contacter/pull/4#discussion_r1412853554>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXT6LQO27QEVMSSFTGMNQSLYHN2SHAVCNFSM6AAAAAA7ICSTSCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTONRQHEZTSMRRGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

vpyzhyanov commented on src/main/java/ru/anykeyers/factories/RepositoryFactory.java:
```diff
@@ -0,0 +1,70 @@
+
+    private final ApplicationProperties applicationProperties;
+
+    private final StorageType STORAGE_TYPE;
```
имя написано как для константы, но это не константа. Пожалуйста, не надо вводить никого в заблуждение, измените на camelCase

Krayseer replied:
изменим

Krayseer commented:
Исправлено в новой версии

сб, 2 дек. 2023 г. в 23:55, Пыжьянов Вячеслав ***@***.***>:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/main/java/ru/anykeyers/commands/Command.java
> <https://github.com/Krayseer/contacter/pull/4#discussion_r1412853554>:
>
> > +    }
> +
> +    public boolean isNeedAuthenticate() {
> +        return isNeedAuthenticate;
> +    }
> +
> +    public static Command getCommandByValue(String value) {
> +        for (Command command : Command.values()) {
> +            if (command.getCommandValue().equals(value)) {
> +                return command;
> +            }
> +        }
> +        return null;
> +    }
> +
> +    public static String getAllCommands() {
>
> не исправлено!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Krayseer/contacter/pull/4#discussion_r1412853554>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXT6LQO27QEVMSSFTGMNQSLYHN2SHAVCNFSM6AAAAAA7ICSTSCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTONRQHEZTSMRRGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

vpyzhyanov commented on src/main/java/ru/anykeyers/factories/RepositoryFactory.java:
```diff
@@ -0,0 +1,70 @@
+
+/**
+ * Фабрика для создания репозиториев<br/>
+ * Типы репозиториев устанавливаются в application.properties в ключе STORAGE_TYPE
```
это не правда, ключ называется не так. `"storage.type"`

Krayseer replied:
не заметил, исправим

vpyzhyanov commented on src/main/java/ru/anykeyers/factories/RepositoryFactory.java:
```diff
@@ -0,0 +1,70 @@
+            case FILE:
+                return new FileUserRepository(applicationProperties.getSetting("file-users-table"));
+            default:
+                return null;
```
почему null? Хотите получать NullPointerException в совершенно неожиданном месте, вместо нормального человекочитаемого сообщения с подробным стеком ошибки?

В общем тут надо бросать исключение и читаемым сообщением об ошибке

Krayseer replied:
сделаем

vpyzhyanov commented on src/main/java/ru/anykeyers/factories/RepositoryFactory.java:
```diff
@@ -0,0 +1,70 @@
+        FILE,
+        POSTGRES,
+        MY_SQL,
+        CASSANDRA
```
Все неиспользуемые значения должны быть безжалостно удалены

Krayseer replied:
в новой версии удалены

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/BaseFileRepository.java:
```diff
@@ -0,0 +1,44 @@
+        try {
+            FileUtils.writeLines(dbFile, "UTF-8", data, false);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
тут должно быть человекочитаемое сообщение

Krayseer replied:
сделаем

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/FileUserRepository.java:
```diff
@@ -0,0 +1,54 @@
+        if (infoByUsername.get(user.getUsername()) == null) {
+            infoByUsername.put(user.getUsername(), new UserData());
+        }
+        UserData userInfo = (UserData) infoByUsername.get(user.getUsername());
```
Никаких кастов!

Krayseer replied:
в обновленной версии все исправлено

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/FileUserRepository.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public boolean existsByUsername(String username) {
+        return getInfosByUsername().containsKey(username);
```
ого, это получается вы при каждом обращении читаете весь файл заново! Не надо так издеваться над процессором и жёстким диском. Сохраняйте их в память, а в БД лазьте только при изменении значений.

Krayseer replied:
данное решение было принято из за того, что раньше мы по одному username могли работать и в консоли и в телеграмме, для них были разные userRepository, так как для разных приложений разные экземпляры. В классе мы с вами обсудили что лучше сделать составной ключ из username и типа бота. Теперь в новой версии как раз сохраняется в память

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/BaseFileRepository.java:
```diff
@@ -0,0 +1,44 @@
+     * Получить карту вида [имя пользователя -> данные по конкретному типу]
+     */
+    protected Map<String, ObjectData<T>> getInfosByUsername() {
+        return formatter.initFromFile(dbFile);
```
Вижу логическую неувязочку:

1. Мы находимся в классе, который управляет данными, в частности **работает с файлом**
2. Метод, который должен **прочитать данные из файла** отдаёт эту задачу кому-то другому
3. Formatter - это класс, который решает единственную задачу - **преобразует данные в текстовый вид**. Он обычно **используется для вывода информации**, а не для чтения.

Вы перешли на тёмную сторону. Вы исключены из совета джедаев)

Krayseer replied:
Данный метод убран из этого класса в новой версии

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/FileUserRepository.java:
```diff
@@ -0,0 +1,54 @@
+
+    public FileUserRepository(String userFilePath) {
+        super(userFilePath);
+        formatter = new UserFormatter();
```
Инициализация protected поля? Не ребята, так дело не пойдёт. Инициализация должна быть в конструкторе. Очень легко забыть проинициализировать поле, когда будете добавлять ещё одну реализацию.

Krayseer replied:
в обновленной версии уже исправлено

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/BaseFileRepository.java:
```diff
@@ -0,0 +1,44 @@
+ */
+public abstract class BaseFileRepository<T> {
+
+    protected ObjectFormatter<T> formatter;
```
За такую грубую ошибку с инициализацией поля пишу замечание: нарушена инкапсуляция. Поле должно быть private

Krayseer replied:
в обновленной версии уже исправлено

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/formatters/ObjectFormatter.java:
```diff
@@ -0,0 +1,27 @@
+ * Интерфейс форматтера, описывающий поведение для обработки файловых БД
+ * @param <T>
+ */
+public interface ObjectFormatter<T> {
```
1. Ранее я писал что все понимают под форматером. Получается имя класса и описание говорят абсолютно о разных вещах.
2. Я несколько раз прочитал это описание, но до сих пор ничего не понял
3. Судя по методам это что-то вроде Mapper'а, т.е. класса, который преобразует один тип в другой и наоборот

Ну и ранее я писал, что он выполняет чужую функцию. Размазали вы одну ответственность между классами и создали коллективную безответственность

Krayseer replied:
Исправили в новой версии, теперь это Parser и делает он только то, что парсит строку и объект и наоборот

vpyzhyanov replied:
ну я же написал, что все это называют `Mapper`! Назовите его общепринятым именем. 

Parser - это класс, которые строку преобразует в объект. То есть Mapper = Parser + Formatter.

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/formatters/ObjectFormatter.java:
```diff
@@ -0,0 +1,27 @@
+     * Инициализация карты вида [имя пользователя -> данные пользователя]
+     * @param file файловая БД
+     */
+    Map<String, ObjectData<T>> initFromFile(File file);
```
тут можно увидеть, что функциональность связана с файлом, но в названии нет про это ни слова. Вы явно это пытались скрыть, но от меня не скроешь))

Метод с таким смыслом обычно называется `parse`

Krayseer replied:
ObjectFormatter был вложен в пакет file, указывая на то, что данный форматтер используется для файловых БД
initFromFile - инициализация карты с данными из файла

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/formatters/UserFormatter.java:
```diff
@@ -0,0 +1,96 @@
+    public Map<String, ObjectData<User>> initFromFile(File file) {
+        Map<String, ObjectData<User>> collectionMap = new HashMap<>();
+        try {
+            List<String> lines = FileUtils.readLines(file, "UTF-8");
```
вот тут он лезет в чужую ответственность

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/data/ObjectData.java:
```diff
@@ -0,0 +1,30 @@
+    /**
+     * Получить информацию об объекте
+     */
+    Object getData();
```
Никаких Object. Типизация в Java - её сильная сторона, не стоит от неё отказываться

Krayseer replied:
после обсуждения в аудитории было убрано

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/data/GroupData.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Информация о группах
+ */
+public class GroupData implements ObjectData<Group> {
```
Все эти классы устроены абсолютно одинаково, значит можно обойтись одним классом, вместо четырёх. А пользователя тоже можно в коллекцию добавить, просто там будет один элемент.

Krayseer replied:
в обновленной версии уже исправлено, данных классов вообще не будет

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/data/ObjectData.java:
```diff
@@ -0,0 +1,30 @@
+     * Удалить данные об объекте
+     * @param data данные для удаления
+     */
+    void removeData(T data);
```
Нигде не используется

Krayseer replied:
в обновленной версии уже исправлено, данных классов вообще не будет

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/data/ContactData.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Override
+    public void addData(Contact contact) {
+        contacts.removeIf(currentContact -> Objects.equals(currentContact.getId(), contact.getId()));
```
что-то вы плохо учили коллекции, можно сделать просто remove, так будет намного эффективнее. Правда для этого сравнение должно быть именно по id, а не как сейчас у вас.

Вот значит как вы сделали редактирование. Не самый очевидный ход.

Krayseer replied:
сохранение пользователя в Set с абсолютно одинаковыми хэшами делали так, что в Set были два одинаковых пользователя, решили это вот таким способом

vpyzhyanov commented on src/main/java/ru/anykeyers/domain/Contact.java:
```diff
@@ -0,0 +1,87 @@
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof Contact contact)) return false;
+        return Objects.equals(name, contact.getName()) && Objects.equals(phoneNumber, contact.getPhoneNumber());
```
реализация некорректна

Krayseer replied:
решили что для классов не обязательно сравнение по всем полям, так как username и названия группы вполне достаточно, но исправим

vpyzhyanov replied:
Не исправлено. По прежнему некорректно

vpyzhyanov commented on src/main/java/ru/anykeyers/domain/Group.java:
```diff
@@ -0,0 +1,87 @@
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof Group group)) return false;
+        return Objects.equals(username, group.getUsername()) && Objects.equals(name, group.getName()) && Objects.equals(contacts, group.getContacts());
```
у всех остальных тоже некорректная реализация

Krayseer replied:
решили что для классов не обязательно сравнение по всем полям, так как username и названия группы вполне достаточно, но исправим

vpyzhyanov replied:
Не исправлено!

vpyzhyanov replied:
Замечание по прежнему не исправлено!!! Даже во второй задаче!

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/data/GroupData.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    @Override
+    public void addData(Group group) {
```
вот смотрю на этот класс, да и на весь `ObjectData`, но не очень понимаю зачем он вам вообще нужен. Вы зачем-то создали класс, для того, чтобы работать с коллекциями объектов, хотя есть же **коллекции**. Методы добавления и удаления там есть, зачем дублировать логику системных классов?

Krayseer replied:
Исправлено в новой версии, вообще был расчёт на то, что в данных может быть или объект или коллекция объектов, но теперь просто делается Collections.singletonList

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/formatters/UserFormatter.java:
```diff
@@ -0,0 +1,96 @@
+                collectionMap.computeIfAbsent(username, k -> new UserData()).addData(user);
+            });
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
должно быть читаемое сообщение об ошибке

Krayseer replied:
Добавлено в новой версии

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/BaseFileRepository.java:
```diff
@@ -0,0 +1,44 @@
+ * Базовый класс для файловых репозиториев
+ * @param <T> класс сущности
+ */
+public abstract class BaseFileRepository<T> {
```
Теперь я понял, что этот класс вовсе не является репозиторием. Он не предоставляет доступ к данным. Его задача читать и записывать файл. А метода `getInfosByUsername` тут вообще быть не должно.

Следовательно вы некорректно использовали наследование и наследовались от сущности, с которой нет ничего родственного. Тут надо использовать делегирование.

Krayseer replied:
правильно ли я понял что нужно убрать этот класс и просто в каждом файловом репозитории инициализировать карту типа String -> Collection<T>

Krayseer replied:
понял в чем дело, исправим

vpyzhyanov commented on src/main/java/ru/anykeyers/factories/StateProcessorFactory.java:
```diff
@@ -0,0 +1,42 @@
+     * @param stateType тип состояния(контакты, группа...)
+     * @return обработчик
+     */
+    public Processable getStateProcessorByType(StateType stateType) {
```
`StateProcessorFactory` должен создавать объекты `StateProcessor`. Здесь же какой-то непонятный тип. Похоже проблема в именовании, нужно переименовать в  `StateProcessor`

Krayseer replied:
Исправили на StateProcessor

vpyzhyanov commented on src/main/java/ru/anykeyers/processors/states/BaseStateProcessor.java:
```diff
@@ -0,0 +1,61 @@
+    /**
+     * Карта вида [состояние -> его обработчик]
+     */
+    protected final Map<State, StateHandler> stateHandlers;
```
нарушения инкапсуляции во всём файле

Krayseer replied:
исправлено в новой версии, теперь private

vpyzhyanov commented on src/main/java/ru/anykeyers/processors/states/BaseStateProcessor.java:
```diff
@@ -0,0 +1,61 @@
+    /**
+     * Регистрация обработчиков состояний {@link #stateHandlers}
+     */
+    public abstract void registerStatesHandlers();
```
1. Метод вызывается только в конструкторе текущего класса, тогда с какой стати ему быть **public**?
2. Метод вообще лишний т.к. наследники сами могут без проблем добавить обработчики в **своём** конструторе. Правда когда вы решите проблему с инкапсуляцией, вам нужно будет добавить метод `registerHandler`

Krayseer replied:
в новой версии сделали по вашему совету

vpyzhyanov commented on src/main/java/ru/anykeyers/processors/states/BaseStateProcessor.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    public String processState(User user, String message) {
+        StateHandler stateHandler = stateHandlers.get(user.getState());
+        return stateHandler.handleState(user, message);
```
здравствуй NPE, порадуй меня своей неочевидностью происходящего)))

Krayseer replied:
в новой версии добавили две проверки на null: user.getState() и stateHandlers.get(user.getState())

vpyzhyanov reviewed (changes requested):
Вы перешли на тёмную сторону. Вы исключены из совета джедаев))

vpyzhyanov commented on src/main/resources/application.properties:
```diff
@@ -1,6 +1,11 @@
+console.username=evgeshanyaa
+
+telegram.username=contacter_app_bot
+telegram.token=6355701319:AAHLw3nBkLmEmP7W7DS1ehALdDpIEO4n9bs
```
Нарушение безопасности: токен в открытом виде, получайте -1 балл за гит.

Krayseer replied:
дак вы же сказали чтобы проект работал когда вы будете его запускать у себя, мы могли его закинуть в .env файл

Krayseer replied:
исправим

vpyzhyanov replied:
Но при этом есть требование, что ключей и паролей в репозитории быть не должно

vpyzhyanov commented on src/test/java/ru/anykeyers/processors/states/impl/ContactStateProcessorTest.java:
```diff
@@ -0,0 +1,124 @@
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.*;
+
+// TODO: 29.11.2023 Нужно доделать 
```
нужно доделать

Krayseer replied:
доделали в новой версии

vpyzhyanov commented on src/test/java/ru/anykeyers/processors/states/impl/ContactStateProcessorTest.java:
```diff
@@ -0,0 +1,124 @@
+//    }
+//
+//    @Test
+//    public void testProcessState_InvalidState() {
```
Закомментированному коду не место в проекте

Krayseer committed:
Большие обновления по архитектуре
1) Сервисы вынесены в интерфейсы и их реализации
2) Убраны классы UserData, ContactData, GroupData из за их ненадобности - заменены на коллекции данных
3) Исправлен BaseStateProcessor, убран абстрактный метод для регистрации обработчиков, добавлен метод который регистрирует команды (поправлена инкапсуляция)
4) Из интерфейса Bot убраны методы receiveMessage и start
5) Убраны статические методы из класса Command
6) Добавлен MessageProcessor и из него уже идет делегирование обработки сообщения либо в CommandProcessor, либо в StateHandlerProcessor
7) Поправлены методы equals и hashCode у классов User, Contact, Group
8) В ContactRepository добавлен метод findByUsernameAndId
9) Исправлен ContactStateProcessor по новым стандартам BaseStateProcessor
10) Исправлены классы FileObjectFormatter, теперь они реализуют FileObjectParser, в методах которого идет парсинг из объекта в строку или строки в объекта
11) Инициализация данных из файла и сохранение данных в файловую БД теперь происходит в сервисе FileService
12) Processable переименован в StateProcessor
13) Во все обработки ошибок добавлены человекочитаемые сообщения
14) Добавлены тесты для сервисов, процессоров и репозиториев(в том числе парсеров и сервиса файлов)

Krayseer committed:
Настройки приложения убраны в .gitignore

vpyzhyanov commented on src/main/java/ru/anykeyers/contexts/Messages.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Сообщения приложения
+ */
+public class Messages {
```
Не надо было изобретать велосипед, надо было воспользоваться уже готовым классом для локализации `java.util.ResourceBundle` (хоть и для одной локали)

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/parsers/FileObjectParser.java:
```diff
@@ -0,0 +1,22 @@
+     * @param object объект для записи в БД
+     * @return строка соответствующего вида для сохранения в файловую БД
+     */
+    String parseTo(T object);
```
некорректное имя метод. Вроде и JavaDoc написали, но он почему-то вам не помог, странно. Должно быть `format`

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/parsers/FileObjectParser.java:
```diff
@@ -0,0 +1,22 @@
+     * @param line строка
+     * @return T объект, сформированный из строки
+     */
+    T parseFrom(String line);
```
Должно быть просто `parse`

vpyzhyanov reviewed (changes requested):
Исправляйте последние замечания и можете мёрджить. Задача принята

Avrooid committed:
Исправил equals и hashCode у Contact, Group.
Переименовал форматтеры в Mapper с методами format, parse
Классы Messages и ApplicationProperties теперь используют ResourcesBundle

vpyzhyanov commented:
Вторая задача приниматься не будет пока эта не смёржена
